### PR TITLE
Create torus_z.py

### PR DESCRIPTION
**Pull request by Talen-Ayers**
Added the torus surface to this new fork of the dev branch in order to create a draft pull request for the project.

### CODE REVIEW

**Comment by ilhamv** [comment]
Thanks for working on this, @Talen-Ayers !

File: mcdc/transport/geometry/surface/torus_x.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
+"""
+Torus: Implicit equation of a torus radially symmetric about the z-axis in the cartesian plane
+
+r^2 = f(x, y, z) = ( sqrt( (x - A)^2 + (y - B)^2) - R )^2 + (z - C)^2
```
Comment (ilhamv) [inline mcdc/transport/geometry/surface/torus_x.py]: Should we move the r^2 to the right-hand side to get the conventional behavior of a surface equation? That is:

-  f(x,y,z) = 0 --> (x,y,z) is on the torus
-  f(x,y,z) < 0 --> (x,y,z) is inside the torus
-  f(x,y,z) > 0 --> (x,y,z) is outside the torus

File: mcdc/transport/geometry/surface/torus_z.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    #This is the same issue that the reflection and normal component functions had, if the particle is outside the torus space
+    if (r**2 - ( math.sqrt( (x - A)**2 + (y - B)**2) - R )**2) < 0:
+        return INF
+
+    # Check if the particle is above or below the centerline of the torus and use the appropriate sign for the square root
+    return (
+        (C + math.sqrt( r**2 - ( math.sqrt( (x - A)**2 + (y - B)**2) - R )**2 ) - z)
+        if z >= C else
+        -(C - math.sqrt( r**2 - ( math.sqrt( (x - A)**2 + (y - B)**2) - R )**2 ) - z)
+    )
```
Comment (ilhamv) [inline mcdc/transport/geometry/surface/torus_z.py:56]: Per the standardized f(x,y,z) formulation (in the previous comment), the function `evaluate` really just evaluates the function f(x,y,z) given the particle location (x,y,z).

Note that we return INF in `get_distance` when we have a particle moving away from, or not going to intersect with, the surface.

File: mcdc/transport/geometry/surface/torus_x.py (hunk 1 above)
Comment (ilhamv) [inline mcdc/transport/geometry/surface/torus_x.py]: Should this be the formula for torus-z?

File: mcdc/transport/geometry/surface/torus_z.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        if isinstance(solution, complex):
+            pass
+        elif solution >= 0:
+            real_roots.append(solution)
+
+    if len(real_roots) == 0: # Ending the calculation if there are no valid solutions
+        return INF
+
+    # Using the smallest root to get the value of t at the first point of intersection
+    # If the direction vector is normalized, the distance to intersection is just the value of t
```
Comment (ilhamv) [inline mcdc/transport/geometry/surface/torus_z.py]: The direction vector is always normalized.

File: mcdc/transport/geometry/surface/torus_z.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    print("Derivitive Magnitude: ", norm)
+    print("Derivitives: ", dx, dy, dz)
+    print("C: ", c)
+    print("Normals: ", nx, ny, nz)
+    print(particle["ux"], particle["uy"], particle["uz"])
+
+# Due to the sqrt in the derivative functions, if the particle is not in the x-y space where the torus has z values, it will result in complex numbers
+# Specificaly the term: math.sqrt(r**2 - (R - inv_root)**2), where inverse root = math.sqrt((A-x)**2 + (B-y)**2)
+@njit
+def get_normal_component(particle_container, surface):
```
Comment (ilhamv) [inline mcdc/transport/geometry/surface/torus_z.py:103]: Note that, whenever this function is called during the transport, the particle position (x,y,z) would yield
f(x,y,z) = 0.0 +- COINCIDENCE_TOLERANCE (the particle is on the surface, or evaluate ~ 0.0)

File: mcdc/transport/geometry/surface/torus_z.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        (C + math.sqrt( r**2 - ( math.sqrt( (x - A)**2 + (y - B)**2) - R )**2 ) - z)
+        if z >= C else
+        -(C - math.sqrt( r**2 - ( math.sqrt( (x - A)**2 + (y - B)**2) - R )**2 ) - z)
+    )
+
+
+# Due to the sqrt in the derivative functions, if the particle is not in the x-y space where the torus has z values, it will result in complex numbers
+# Specificaly the term: math.sqrt(r**2 - (R - inv_root)**2), where inverse root = math.sqrt((A-x)**2 + (B-y)**2)
+@njit
+def reflect(particle_container, surface):
```
Comment (ilhamv) [inline mcdc/transport/geometry/surface/torus_z.py:60]: I think it is rare to have a reflecting torus boundary condition. But it would be cool to have it when one needs it!

**Comment by ilhamv** [comment]
I think the next step would be setting up the unit test. You can check [this](https://github.com/CEMeNT-PSAAP/MCDC/blob/dev/test/unit/transport/geometry/surface/plane_x.py) out as a reference.

To manually run the test, we do `pytest <the file>` on the directory the file is in.
To automatically run all the unit tests, we execute this [script](https://github.com/CEMeNT-PSAAP/MCDC/blob/dev/test/unit/run.py).

===

As you can see, now there are only tests for axis-aligned planes. Adding unit tests for the other surfaces for warming up would be great if you are interested! 😉

**Comment by ilhamv** [comment]
> I think the next step would be setting up the unit test. You can check [this](https://github.com/CEMeNT-PSAAP/MCDC/blob/dev/test/unit/transport/geometry/surface/plane_x.py) out as a reference.

Actually, there is another piece that you need to work on first before setting up the unit test. And that is the torus surface creation. Note that:
- `mcdc/object_/surface.py` describes how we store the surface object data (what is still missing now for torus)
- `mcdc/transport/geometry/surface/*py` are the functions that use the surface object data during the transport simulation (what you just added)

In `mcdc/object_/surface.py`, the torus surface would need to be added to the "Type-based creation methods" section and the `__repr__` function method for nice-printing.

Note that in the creation method, we store the coefficients in A, B, ..., I, J. This means, you would need to re-letter the coefficients that you expect in your functions; for example, changing "r" and "R" into "I" and "J".

File: mcdc/object_/surface.py [hunk 6]
```diff
@@ -585,6 +590,49 @@ def Quadric(
...
+        A: float = 0.0,
+        B: float = 0.0,
+        C: float = 0.0,
+        R: float = 0.0,
+        r: float = 0.0,
+        boundary_condition: str = "none",
+    ):
+        """
+        Create a torus on the x-y plane radially symetric around the z axis:
+            f(x, y, z) = ( sqrt[(x - A)^2 + (y - B)^2] - R )^2 + (z - C)^2 - r^2
```
Comment (melekderman) [inline mcdc/object_/surface.py:843]: I'm not sure whether this formulation represents `torus_x` or `torus_z`?

File: mcdc/transport/geometry/surface/torus_z.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
+"""
+Torus: Implicit equation of a torus radially symmetric about the z-axis in the cartesian plane
+
+f(x, y, z) = ( sqrt[(x - A)^2 + (y - B)^2] - R )^2 + (z - C)^2 - r^2
```
Comment (melekderman) [inline mcdc/transport/geometry/surface/torus_z.py:4]: I think it should be `torus_z`, as mentioned in my comment in `surface.py`.

File: mcdc/transport/geometry/surface/torus_x.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
+"""
+Torus: Implicit equation of a torus radially symmetric about the z-axis in the cartesian plane
+
+f(x, y, z) = ( sqrt[(x - A)^2 + (y - B)^2] - R )^2 + (z - C)^2 - r^2
+
+Where R is the radius of the shape as a whole, and r is the radius of the circle that is revolved to create the donut
+
+Removing the square roots leaves you with the following equation:
+f (x, y, z) = ( x^2 + y^2 + z^2 + R^2 - a^2 )^2 - 4R^2 * (x^2 + y^2)
```
Comment (melekderman) [inline mcdc/transport/geometry/surface/torus_x.py]: f (x, y, z) = ( x^2 + y^2 + z^2 + R^2 - `r^2` )^2 - 4R^2 * (x^2 + y^2)

File: mcdc/transport/geometry/surface/torus_z.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+    # Dot products that come up frequently in the torus-ray intersection equation
+    G = ux*ux + uy*uy + uz*uz
+    H = 2.0 * (x*ux + y*uy + z*uz)
+    I = x*x + y*y + z*z
+
+    J = ux*ux + uy*uy
+    K = 2.0 * (x*ux + y*uy)
+    L = x*x + y*y
+
```
Comment (melekderman) [inline mcdc/transport/geometry/surface/torus_z.py:182]: It seems like the COINCIDENCE_TOLERANCE check is missing?

File: mcdc/transport/geometry/surface/torus_x.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    ny = dy / norm
+    nz = dz / norm
+
+    # Reflect
+    c = 2.0 * (nx * ux + ny * uy + nz * uz) # Magnitutde component of the projection of the particle onto the surface normal
+    particle["ux"] -= c * nx
+    particle["uy"] -= c * ny
+    particle["uz"] -= c * nz
+
+    print("Directions: ", ux, uy, uz)
```
Comment (melekderman) [inline mcdc/transport/geometry/surface/torus_x.py]: REMINDER: Debug prints should be deleted before merging the PR.

**Review by melekderman — COMMENTED** [review]
- Comments for PR #364

**Review by melekderman — APPROVED** [review]
Hi @Talen-Ayers,

Thanks again for the great work on the torus_z surface implementation!

I’ve just gone through the PR and made some final adjustments/fixes to get everything across the finish line:

- I replaced `poly.polyroots` with `np.roots` to ensure the code is Numba-compatible, which is important for the performance.
- I resolved a few type errors and unit test failures that were popping up.
- I also added some TODOs regarding near-tangent cases for future improvements - if needed.

The CI checks are looking good now, so we are ready to go! I'll be merging this into dev soon.

Thank you for your first contribution!
